File: 2019/Advent of Code 2019/Days/Day1.cs
using System.Collections.Generic;
using System.Linq;
using AoC2019.Extensions;

namespace AoC2019.Days
{
    public sealed class Day1 : DayBase<IEnumerable<int>, int>
    {
        public override string Name => "Day 1: The Tyranny of the Rocket Equation";

        public override IEnumerable<int> ParseInput(IEnumerable<string> lines)
        {
            return lines.Select(int.Parse);
        }

        public override int RunPart1(IEnumerable<int> input)
        {
            return input.Sum(Fuel);
        }

        public override int RunPart2(IEnumerable<int> input)
        {
            return input.Sum(m => Fuel(m).Sequence(Fuel).TakeWhile(f => f > 0).Sum());
        }

        private static int Fuel(int moduleMass) => moduleMass / 3 - 2;
    }
}

File: 2021/src/main/kotlin/de/thesmoun/aoc/days/Day1.kt
package de.thesmoun.aoc.days

class Day1 : Day<Collection<Int>, Int>("Day 1: Sonar Sweep") {

    override fun parseInput(input: Collection<String>) = input.map { it.toInt() }

    override fun runPart1(input: Collection<Int>): Int = input.windowed(2).count { it[0] < it[1] }

    override fun runPart2(input: Collection<Int>): Int = input.windowed(3)
        .windowed(2).count { it[0].sum() < it[1].sum() }
}
